File: ThreadBacktrace/Source/Help/BacktraceHelp.swift
//
//  BacktraceHelp.swift
//  ThreadBacktrace
//
//  Created by yans on 2019/9/5.
//

import Foundation
import Darwin

//MARK: 线程相关 私有
/// 获取线程调用堆栈
@_silgen_name("mach_backtrace")
public func backtrace(_ thread: thread_t,
                      stack: UnsafeMutablePointer<UnsafeMutableRawPointer?>,
                      maxSymbols: Int32) -> Int32

//MARK: extension
extension Character {
    var isAscii: Bool {
        return unicodeScalars.allSatisfy { $0.isASCII }
    }
    var ascii: UInt32? {
        return isAscii ? unicodeScalars.first?.value : nil
    }
}

extension String {
    var ascii : [Int8] {
        var unicodeValues = [Int8]()
        for code in unicodeScalars {
            unicodeValues.append(Int8(code.value))
        }
        return unicodeValues
    }
}

File: ThreadBacktrace/Source/Help/backtrace_help.c
//
//  backtrace.c
//  ThreadBacktrace
//
//  Created by yans on 2019/9/5.
//

#include "backtrace_help.h"

/// 主线程 ID 利用 C 构造函数设置
static mach_port_t _main_thread_id;

__attribute__((constructor)) static
void _setup_main_thread() {
    _main_thread_id = mach_thread_self();
}

/// 获取主线程 ID 
mach_port_t get_mach_main_thread() {
    return _main_thread_id;
}

bool has_dli_fname(struct dl_info info) {
    return info.dli_fname != NULL;
}
